### PR TITLE
Add `RandomYield` to `itertools`.

### PR DESCRIPTION
**Pull request by huibinshen**
*Issue #, if available:*

*Description of changes:*

Given a list of iterators, generate samples uniformly or according to the probabilities if given.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by huibinshen** [comment]
cc @abdulfatir

File: src/gluonts/itertools.py [hunk 1]
```diff
@@ -289,6 +289,43 @@ def __iter__(self):
...
+            probs = [
+                prob / sum(self.probabilities) for prob in self.probabilities
+            ]
+            idx = self.random_state.choice(range(len(self.iterators)), p=probs)
+
+            try:
+                data = next(self.iterators[idx])
+                success = True
+            except StopIteration:
+                self.probabilities[idx] = 0.0
```
Comment (jaheba) [inline src/gluonts/itertools.py]: I don't think this will work, as the probabilities have to sum up to 1.

File: src/gluonts/itertools.py [hunk 2]
```diff
@@ -289,6 +289,43 @@ def __iter__(self):
...
+@dataclass
+class MixIterators:
+    iterators: List[Iterator]
+    probabilities: Optional[List[float]] = None
+    random_state: np.random.RandomState = np.random.RandomState()
+
+    def __post_init__(self):
+        if self.probabilities:
+            # copy the probabilities as we may modify it
+            self.probabilities = [prob for prob in self.probabilities]
```
Comment (jaheba) [inline src/gluonts/itertools.py]:
```suggestion
            self.probabilities = list(self.probabilities)
```

**Review by jaheba — COMMENTED** [review]
There are some things which are not clear to me. The idea here is to sample from multiple underlying iterators, given some probability.

However, the question is what should happen if one source gets exhausted. I guess there are two approaches: Either stop yielding values entirely, or only yielding from the remaining sources. I can see the argument for both.

Also, should it be possible to rerun the iterator? E.g, if the input are lists then we could get different permutations each time we execute this.

File: src/gluonts/itertools.py [hunk 3]
```diff
@@ -289,6 +289,43 @@ def __iter__(self):
         return filter(self.fn, self.iterable)
 
 
+@dataclass
+class MixIterators:
+    iterators: List[Iterator]
```
Comment (jaheba) [inline src/gluonts/itertools.py]: I think I would make this
```suggestion
    iterators: List[Iterable]
```

File: src/gluonts/itertools.py [hunk 4]
```diff
@@ -289,6 +289,43 @@ def __iter__(self):
...
+    probabilities: Optional[List[float]] = None
+    random_state: np.random.RandomState = np.random.RandomState()
```
Comment (jaheba) [inline src/gluonts/itertools.py]: I think you would want to use `field` with `default_factory` here, that way each instance gets its own state.

File: src/gluonts/itertools.py [hunk 5]
```diff
@@ -289,6 +289,43 @@ def __iter__(self):
...
+    def __post_init__(self):
+        if self.probabilities:
+            # copy the probabilities as we may modify it
+            self.probabilities = [prob for prob in self.probabilities]
+        else:
+            self.probabilities = [1.0 / len(self.iterators)] * len(
+                self.iterators
+            )
+
+    def __iter__(self):
```
Comment (jaheba) [inline src/gluonts/itertools.py:327]: I think we can just do everything in `__iter__`, no need for `__next__`.

File: src/gluonts/itertools.py (hunk 1 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: It normalizes the probabilities in the beginning of the loop.

File: src/gluonts/itertools.py (hunk 2 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: Will do.

File: src/gluonts/itertools.py [hunk 6]
```diff
@@ -289,6 +289,43 @@ def __iter__(self):
         return filter(self.fn, self.iterable)
 
 
+@dataclass
+class MixIterators:
```
Comment (abdulfatir) [inline src/gluonts/itertools.py]: I don't like the name `MixIterators`.

File: src/gluonts/itertools.py (hunk 3 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: I talked with Lorenzo specifically on this. If we make it `Iterable`, we will need to create `Iterator`s inside the class, which will be hidden from user. To be explicit, we thought `Iterator` might be better.

**Comment by jaheba** [comment]
This should work:


```py
    def __iter__(self):
        iterators = [iter(it) for it in self.iterables]
        indices = range(len(iterators))
        p = list(self.p)

        while True:
            idx = self.random_state.choice(indices, p=p)

            try:
                yield next(iterators[idx])
            except StopIteration:
                p[idx] = 0.0
                if sum(p) == 0.0:
                    return
                p /= sum(p)
```

File: src/gluonts/itertools.py (hunk 4 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: Will take a look!

File: src/gluonts/itertools.py (hunk 6 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: Any suggestion? `SampleFromIterators` or `MetaIterators`?

File: src/gluonts/itertools.py (hunk 5 above)
Comment (huibinshen) [reply src/gluonts/itertools.py:327]: This is related to the `Iterable` or `Iterator` question. If we stick to `Iterator`, then I think it makes sense to implement `__next__` method. So you pass in a list of `Iterator`s, you get an `Iterator` out.

File: src/gluonts/itertools.py (hunk 3 above)
Comment (jaheba) [reply src/gluonts/itertools.py]: See my proposal below, I think this should be iterators which we then call `iter` on.

File: src/gluonts/itertools.py (hunk 6 above)
Comment (jaheba) [reply src/gluonts/itertools.py]: Maybe something with "interleave"?

Comment (abdulfatir) [reply src/gluonts/itertools.py]: How about `WeightedIterables` or something with probabilistic?

**Comment by huibinshen** [comment]
> This should work:
> 
> ```python
>     def __iter__(self):
>         iterators = [iter(it) for it in self.iterables]
>         indices = range(len(iterators))
>         p = list(self.p)
> 
>         while True:
>             idx = self.random_state.choice(indices, p=p)
> 
>             try:
>                 yield next(iterators[idx])
>             except StopIteration:
>                 p[idx] = 0.0
>                 if sum(p) == 0.0:
>                     return
>                 p /= sum(p)
> ```

Because one creates an `Iterator` in the `__iter__`, it always give you the first element in those `Iterator`s. One way is to create these `iterators` as class fields in `__post_init__`. But if we decide to go with `Iterator`, then this would not be needed.

**Comment by abdulfatir** [comment]
> Because one creates an Iterator in the __iter__, it always give you the first element in those Iterators.

Why?

**Comment by huibinshen** [comment]
> > Because one creates an Iterator in the **iter**, it always give you the first element in those Iterators.
> 
> Why?

```
>>> a = list(range(10))
>>> a_it = iter(a)
>>> next(a_it)
0
>>> a_it = iter(a)
>>> next(a_it)
0
>>>
```
No? Or you mean `__iter__` will only be called once so it should be fine?

**Comment by huibinshen** [comment]
> There are some things which are not clear to me. The idea here is to sample from multiple underlying iterators, given some probability.
> 
> However, the question is what should happen if one source gets exhausted. I guess there are two approaches: Either stop yielding values entirely, or only yielding from the remaining sources. I can see the argument for both.

Agree. I will make the documentation clearer.

> Also, should it be possible to rerun the iterator? E.g, if the input are lists then we could get different permutations each time we execute this.

I mean, one could create a new iterator and pass that to the class right?

**Comment by abdulfatir** [comment]
> Or you mean __iter__ will only be called once so it should be fine?

Yes.

File: src/gluonts/itertools.py (hunk 3 above)
Comment (lostella) [reply src/gluonts/itertools.py]: > I think this should be iterators which we then call `iter` on

_iterables_

File: src/gluonts/itertools.py (hunk 6 above)
Comment (jaheba) [reply src/gluonts/itertools.py]: That name confuses me.

**Review by jaheba — CHANGES_REQUESTED** [review]
Let's also add a test as docstring.

File: src/gluonts/itertools.py (hunk 6 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: Changed to `WeightedIterables`.

File: src/gluonts/itertools.py [hunk 7]
```diff
@@ -289,6 +289,65 @@ def __iter__(self):
...
+        iterators = [iter(it) for it in self.iterables]
+        while True:
+            idx = self.random_state.choice(
+                range(len(iterators)), p=self.probabilities
+            )
+
+            try:
+                yield next(iterators[idx])
+            except StopIteration:
+                self.probabilities[idx] = 0
```
Comment (lostella) [inline src/gluonts/itertools.py]: This makes the object stateful: if while iterating the first time some probabilities are set to zero, they will remain zero the second time you iterate. I don't think this is what we want?

I think in this iterable approach we should have a local copy of `self.probabilities` in the scope of `__iter__`

Comment (huibinshen) [reply src/gluonts/itertools.py]: You are absolutely right, changed.

File: src/gluonts/itertools.py [hunk 8]
```diff
@@ -289,6 +289,62 @@ def __iter__(self):
...
+
+    """
+
+    iterables: List[Iterable]
+    probabilities: Optional[List[float]] = None
+    random_state: RandomState = field(default_factory=RandomState)
+
+    def __post_init__(self):
+        if self.probabilities:
+            # copy the probabilities as we may modify it
```
Comment (jaheba) [inline src/gluonts/itertools.py]: This should no longer be needed.

File: src/gluonts/itertools.py [hunk 9]
```diff
@@ -289,6 +289,62 @@ def __iter__(self):
...
+    """
+    Given a list of Iterables `iterables`, generate samples from them.
+
+    When `probabilities` is given, sample iteratbles according to it.
+    When `probabilities` is not given, sample iterables uniformly.
+
+    When one iterable exhausts, the sampling probabilities for it will be set to 0.
+
+        >>> from toolz import take
+        >>> from gluonts.itertools import Cyclic
```
Comment (jaheba) [inline src/gluonts/itertools.py]: No need to import things from the same module.

Comment (jaheba) [reply src/gluonts/itertools.py]: Also, don't think we need it for demonstrating what this does.

File: src/gluonts/itertools.py [hunk 10]
```diff
@@ -289,6 +289,62 @@ def __iter__(self):
         return filter(self.fn, self.iterable)
 
 
+@dataclass
+class WeightedIterables:
```
Comment (jaheba) [inline src/gluonts/itertools.py]: Looking at just the name, I wouldn't know what this does. In some sense this is a scheduler doing resource management over multiple queues. I wish the name would make it clear that we take multiple iterators and fuse it into a single one.

File: src/gluonts/itertools.py [hunk 11]
```diff
@@ -289,6 +289,62 @@ def __iter__(self):
...
+    When `probabilities` is given, sample iteratbles according to it.
+    When `probabilities` is not given, sample iterables uniformly.
+
+    When one iterable exhausts, the sampling probabilities for it will be set to 0.
+
+        >>> from toolz import take
+        >>> from gluonts.itertools import Cyclic
+        >>> a = [1, 2, 3]
+        >>> b = [4, 5, 6]
+        >>> it = iter(WeightedIterables([a, b], [1, 0]))
```
Comment (jaheba) [inline src/gluonts/itertools.py]: I think it's cleaner if you provide the probabilities as a keyword parameter.

File: test/test_itertools.py [hunk 12]
```diff
@@ -277,3 +279,22 @@ def test_join_items():
 
     with pytest.raises(Exception):
         oin_items(left, right, "strict")
+
+
+@pytest.mark.parametrize(
+    "iterables, probabilities, sample_size, samples",
+    [
+        ([[1, 2, 3], [4, 5, 6]], [1.0, 0.0], 5, [1, 2, 3]),
+        ([[1, 2, 3], [4, 5, 6]], [0.0, 1.0], 5, [4, 5, 6]),
```
Comment (jaheba) [inline test/test_itertools.py]: I don't like the test where we set the probability for an input to zero. Couldn't we initialise the random state with some seed?

**Review by jaheba — CHANGES_REQUESTED** [review]

File: test/test_itertools.py (hunk 12 above)
Comment (lostella) [reply test/test_itertools.py]: I think even if we set the seed, newer python or library versions may break the sequence generated by the same seed.

Comment (jaheba) [reply test/test_itertools.py]: No? The rng comes from numpy which should be fairly stable.

Comment (jaheba) [reply test/test_itertools.py]: No? The rng comes from numpy which should be fairly stable.

File: src/gluonts/itertools.py (hunk 9 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: I will remove the `Cyclic` import but leave the `toolz` import.

File: src/gluonts/itertools.py (hunk 10 above)
Comment (huibinshen) [reply src/gluonts/itertools.py]: `WeightedFusionIterables`?

**Review by jaheba — DISMISSED** [review]

**Review by lostella — COMMENTED** [review]
Maybe I’m late to the party, but since names are important I would consider alternatives:
- `YieldRandomlyFrom`
- `YieldRandomly`
- `RandomScheduler`

**Comment by huibinshen** [comment]
> Maybe I’m late to the party, but since names are important I would consider alternatives:
> 
>     * `YieldRandomlyFrom`
> 
>     * `YieldRandomly`
> 
>     * `RandomScheduler`

Yes, you are late :-). 

From the list, I see the key world `Yield` and `Random`. Jasper was proposing `RandomYield`. I thought since we are passing probabilities to the yield, it is nice to have `Probabilistic` in the name, thus `ProbabilisticYield`. But `RandomYield` may also work for me (as it is the largest common factor).

**Review by lostella — APPROVED** [review]

**Pull request merged**